MarsEnv.java

```diff
@@ -147,18 +162,120 @@ void nextSlot() throws Exception {
         }
 
         void moveTowards(int id, int x, int y) throws Exception {
-            Location r1 = getAgPos(id);
-            if (r1.x < x)
-                r1.x++;
-            else if (r1.x > x)
-                r1.x--;
-            else if (r1.y < y)
-                r1.y++;
-            else if (r1.y > y)
-                r1.y--;
-            setAgPos(id, r1);
-            setAgPos(1, getAgPos(1)); // just to draw it in the view
+            Location position = getAgPos(id);
+            if (position.x < x)
+                moveRight(id);
+            else if (position.x > x)
+                moveLeft(id);
+            else if (position.y < y)
+                moveDown(id);
+            else if (position.y > y)
+                moveUp(id);
         }
```

Comment:
isn't there a lot of code duplication among moveLeft(), moveRight(), etc? Why not just use moveTowards() directly?

Reply:
movetowards() (as it's implemented, at least), is used with a target that's farther from the agent, so that it means, take one step towards this far away point (and the environment figures out how to do that). I think we might end up using it as a higher-level plan, if there is such a thing, or a desire.

Reply:
actually it's not clear to me why we're changing movetowards() now, I would eventually move the intelligence to the agent anyway.

Reply:
I have two answers for this. First, I am trying to replace moveTowards() as this currently has all the navigation logic built within it. That's why the moveLeft (soone to be renamed to moveEast) type functions don't use it.

Secondly, the reason why I've been editing the moveTowards() function is that I was using it as a means of testing my new movement functions and environment checks that I've been building.

Reply:
Wouldn't it be more agile to keep moveTowards() as-is, and come up with other functions to do what you want to do? Then eventually you can phase out moveTowards() if you wish, but there's no risk of affecting the behavior of agents that rely on it.